File: musicpimp/src/main/scala/org/musicpimp/local/LocalPlayer.scala
package org.musicpimp.local

import android.content.{Context, Intent}
import android.media.MediaPlayer
import com.mle.util.Scheduling
import java.util.concurrent.ScheduledFuture
import org.musicpimp.PimpApp
import org.musicpimp.audio._
import org.musicpimp.usage.LocalPlayerLimiter
import scala.concurrent.duration._
import scala.util.Try

/**
 * Media player that delegates calls by sending intents to the background
 * audio <code>Service</code>.
 *
 * The constructor is protected as evident from the uncommon syntax.
 *
 * @author mle
 */
class LocalPlayer protected() extends Player with LocalPlaylist {
  override val isLocal = true
  private val DEFAULT_VOLUME = 100
  private var isMuted = false
  private var vol = DEFAULT_VOLUME
  // polls player for time updates
  private var poller: Option[ScheduledFuture[_]] = None
  private var mPlayer: Option[MediaPlayer] = None

  private var playerTrack: Option[Track] = None

  def track = playerTrack

  def track_=(newTrack: Option[Track]) {
    playerTrack = newTrack
    fireEvent(TrackChanged(newTrack))
  }

  def mediaPlayer = mPlayer

  /**
   * Called whenever a track changes. Reinstalls the previous volume & mute state.
   *
   * @param player a player with new a track
   */
  def initMediaPlayer(player: MediaPlayer) {
    mPlayer = Some(player)
    volume(vol)
    mute(isMuted)
  }

  def closeMediaPlayer() {
    mPlayer = None
    fireEvent(PlayStateChanged(PlayStates.NoMedia))
    stopPollingForTimeUpdates()
  }

  def setAndPlay(track: Track) {
    // sets playlist
    set(track)
    // starts background media player
    startIntent(MediaService.RESTART_ACTION)
    //    info(s"Starting playback of: ${track.title}")
  }

  def resume() {
    mediaPlayer.foreach(startOrResume)
  }

  private def startOrResume(mp: MediaPlayer): Unit = {
    mp.start()
    fireEvent(PlayStateChanged(PlayStates.Started))
    startPollingForTimeUpdates()
    track.foreach(t => updateNotification(t, playing = true))
  }

  def pause() {
    val shouldUpdateNotification = isPlaying
    stopPlayback()
    track.foreach(t => {
      if (shouldUpdateNotification) {
        updateNotification(t, playing = false)
      }
    })
  }

  private def cancelNotification(): Unit =
    new Notifications(PimpApp.context).cancel()

  private def updateNotification(track: Track, playing: Boolean): Unit =
    new Notifications(PimpApp.context).displayTrackNotification(track, playing)

  // MediaPlayer.isPlaying may throw IllegalStateException (hahaha)
  private def isPlaying: Boolean =
    mediaPlayer.exists(p => Try(p.isPlaying).filter(_ == true).isSuccess)

  def stop() {
    stopPlayback()
    cancelNotification()
  }

  private def stopPlayback() {
    mediaPlayer.foreach(p => {
      p.pause()
      fireEvent(PlayStateChanged(PlayStates.Stopped))
      stopPollingForTimeUpdates()
    })
  }

  def seek(pos: Duration) {
    mediaPlayer.foreach(_ seekTo pos.toMillis.toInt)
  }

  def playNext() {
    startIntent(MediaService.NEXT_ACTION)
  }

  def playPrevious() {
    startIntent(MediaService.PREV_ACTION)
  }

  def skip(position: Int) {
    index = Some(position)
    startIntent(MediaService.PLAY_PLAYLIST)
  }

  private def startIntent(action: String) {
    val context = PimpApp.context
    val intent = playbackIntent(context, action)
    context.startService(intent)
  }

  private def playbackIntent(ctx: Context, action: String) = {
    val intent = new Intent(ctx, classOf[MediaService])
    intent setAction action
    intent
  }

  def mute(muted: Boolean) {
    mediaPlayer.foreach(player => {
      isMuted = muted
      // ???
      //      if (muted) {
      //        vol = streamVolume
      //      }
      val floatVolume = if (muted) 0.0f else floatify(vol)
      player.setVolume(floatVolume, floatVolume)
      fireEvent(MuteToggled(muted))
    })
  }

  /**
   *
   * @param volume [0, 100]
   */
  def volume(volume: Int) {
    mediaPlayer.foreach(player => {
      vol = volume
      val floatVolume = floatify(volume)
      player.setVolume(floatVolume, floatVolume)
      fireEvent(VolumeChanged(volume))
    })
  }

  /**
   * @return the current volume as an integer within [0, 100]
   */
  //  def streamVolume: Int = {
  //    val am = PimpApp.context.getSystemService(Context.AUDIO_SERVICE).asInstanceOf[AudioManager]
  //    val vol = am.getStreamVolume(AudioManager.STREAM_MUSIC)
  //    val maxVolume = am.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
  //    (100.0 * vol / maxVolume).toInt
  //  }
  // map from [0, 100] to [0.0f, 1.0f]
  def floatify(volume: Int) = 1.0f * volume / 100

  def onPrepared(mp: MediaPlayer): Unit = {
    startOrResume(mp)
  }

  def status: StatusEvent = mediaPlayer.fold(StatusEvent.empty)(player => {
    val state = if (isPlaying) PlayStates.Started else PlayStates.Stopped
    val pos = player.getCurrentPosition.milliseconds
    StatusEvent(currentTrack, state, pos, vol, isMuted, tracks, index)
  })


  private def startPollingForTimeUpdates() {
    val task = Scheduling.every(900 milliseconds) {
      mediaPlayer.foreach(player => {
        fireEvent(TimeUpdated(player.getCurrentPosition.milliseconds))
      })
    }
    poller = Some(task)
  }

  private def stopPollingForTimeUpdates() {
    poller.foreach(_.cancel(true))
    poller = None
  }
}

object LocalPlayer extends LocalPlayer

object LimitedLocalPlayer extends LocalPlayer with LocalPlayerLimiter

File: app/src/main/java/org/musicpimp/media/LocalPlayer.kt
package org.musicpimp.media

import android.content.Context
import android.content.Intent
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.internal.toImmutableList
import org.musicpimp.Duration
import org.musicpimp.Playstate
import org.musicpimp.Track
import org.musicpimp.audio.Player
import org.musicpimp.millis
import org.musicpimp.ui.player.CoverService
import timber.log.Timber

/** Media player that delegates calls by sending intents to the background audio
 * <code>Service</code>. The Service in turn notifies this player of playback updates.
 */
class LocalPlayer(private val appContext: Context, private val covers: CoverService) : Player {
    private val currentTracks: MutableList<Track> = mutableListOf()
    // The player may have a track which is not in the playlist
    var playerTrack: Track? = null
    private val playerTracks = MutableLiveData<Track>()
    private val playlist = MutableLiveData<List<Track>>().apply {
        value = emptyList()
    }
    private val indices = MutableLiveData<Int?>().apply {
        value = null
    }
    private val playstates = MutableLiveData<Playstate>().apply {
        value = Playstate.NoMedia
    }
    private val currentPosition = MutableLiveData<Duration>().apply {
        value = 0.millis
    }
    val tracks: LiveData<Track> = playerTracks
    val list: LiveData<List<Track>> = playlist
    val index: LiveData<Int?> = indices
    val states: LiveData<Playstate> = playstates
    val position: LiveData<Duration> = currentPosition

    private var currentState: Playstate = Playstate.NoMedia

    private val currentIndex: Int?
        get() = index.value

//    val dummy = states.observeForever { s ->
//        playerTrack?.let { t ->
//            updateNotification(t, s)
//        }
//    }

    override fun play(track: Track) {
        currentTracks.clear()
        addAll(listOf(track))
        indices.postValue(0)
        playerTrack = track
        startIntent(PimpMediaService.RESTART_ACTION)
    }

    override fun add(track: Track) {
        addAll(listOf(track))
    }

    override fun addAll(tracks: List<Track>) {
        currentTracks.addAll(tracks)
        playlist.postValue(currentTracks.toImmutableList())
    }

    override fun resume() {
        startIntent(PimpMediaService.RESUME_ACTION)
    }

    override fun next() {
        startIntent(PimpMediaService.NEXT_ACTION)
    }

    override fun prev() {
        startIntent(PimpMediaService.PREV_ACTION)
    }

    override fun skip(idx: Int) {
        indices.postValue(idx)
        startIntent(PimpMediaService.PLAY_PLAYLIST)
    }

    override fun remove(idx: Int) {
        currentTracks.removeAt(idx)
        currentIndex?.let { i ->
            if (idx <= i && i >= 0) {
                indices.postValue(i - 1)
            }
        }
        playlist.postValue(currentTracks.toImmutableList())
    }

    override fun seek(to: Duration) {
        startIntent(PimpMediaService.SEEK_ACTION) { i ->
            i.putExtra(PimpMediaService.POSITION_EXTRA, to)
        }
    }

    override fun stop() {
        startIntent(PimpMediaService.PAUSE_ACTION)
        onState(Playstate.Stopped)
    }

    override fun pause() {
        startIntent(PimpMediaService.PAUSE_ACTION)
        playstates.postValue(Playstate.Paused)
    }

    fun toNext(): Track? {
        val ret = nextTrack()
        ret?.let { _ -> currentIndex?.let { i -> indices.postValue(i + 1) } }
        return ret
    }

    fun toPrev(): Track? {
        val ret = prevTrack()
        ret?.let { _ -> currentIndex?.let { i -> indices.postValue(if (i > 0) i - 1 else 0) } }
        return ret
    }

    private fun nextTrack(): Track? = withIndex { i -> i + 1 }

    private fun prevTrack(): Track? = withIndex { i -> if (i > 0) i - 1 else 0 }

    fun onTrack(track: Track?) {
        playerTrack = track
        playerTracks.postValue(track)
    }

//    private fun updateNotification(track: Track, playstate: Playstate) {
//        GlobalScope.launch {
//            withContext(Dispatchers.IO) {
//                if (playstate == Playstate.Stopped) {
//                    notifications.cancel()
//                } else {
//                    Timber.i("Updating notification with $playstate")
//                    val bitmap = covers.cover(track)
//                    notifications.displayTrackNotification(
//                        track,
//                        playstate == Playstate.Playing,
//                        bitmap
//                    )
//                }
//            }
//        }
//    }

    fun onState(state: Playstate) {
        currentState = state
        playstates.postValue(state)
    }

    fun onPosition(pos: Duration) {
        currentPosition.postValue(pos)
    }

    private fun withIndex(f: (old: Int) -> Int): Track? =
        currentIndex?.let { idx -> trackAt(f(idx)) }

    private fun trackAt(index: Int): Track? =
        if (index < currentTracks.size && index >= 0) currentTracks[index] else null

    private fun startIntent(action: String) {
        startIntent(action) { i -> i }
    }

    private fun startIntent(action: String, prep: (i: Intent) -> Intent) {
        val context = appContext
        val intent = playbackIntent(context, action)
        context.startService(prep(intent))
    }

    private fun playbackIntent(ctx: Context, action: String): Intent {
        val intent = Intent(ctx, PimpMediaService::class.java)
        intent.action = action
        return intent
    }
}
